Enable const tests for bool methods

dollaransh17 opened the pull request:
This change re-enables the const tests for bool::then_some and bool::then that were previously disabled with a FIXME comment referencing issue 110395.

The tests verify that these methods work correctly in const contexts, which is important for compile-time evaluation. These tests check that both the eager evaluation variant (then_some) and lazy evaluation variant (then) can be used to create constant Option values at compile time.

Tests added:
- Const evaluation of false.then_some(0) returning None
- Const evaluation of true.then_some(0) returning Some(0)
- Const evaluation of false.then(zero) returning None
- Const evaluation of true.then(zero) returning Some(0)

<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r? <reviewer name>
-->
<!-- homu-ignore:end -->

dollaransh17 committed:
Make bool::then_some and bool::then const methods

This change makes both `then_some` and `then` methods on bool const-compatible,
allowing them to be used in const contexts for compile-time evaluation.

Benefits:
- Enables compile-time conditional Option creation
- Allows usage in const generics and static initialization
- Provides zero-cost abstraction in const contexts
- Improves ergonomics for const programming patterns

The `then_some` method is straightforward to make const as it only uses
basic conditional logic. The `then` method requires a `~const FnOnce`
bound to ensure the closure can be called in const contexts.

This resolves the FIXME in the test suite and enables the const tests
for these methods that were previously disabled.

rust-log-analyzer commented:
The job **`aarch64-gnu-llvm-20-2`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/18228487817/job/51905831324) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/51905831324) [(plain)](https://github.com/rust-lang/rust/commit/94e80548fa1d1e010690ef97b213ce67cac3b70e/checks/51905831324/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
[RUSTC-TIMING] allocbenches test:true 9.060
[RUSTC-TIMING] corebenches test:true 10.229
[RUSTC-TIMING] alloctests test:true 15.819
[RUSTC-TIMING] alloctests test:true 17.214
error[E0015]: cannot call non-const method `core::bool::<impl bool>::then_some::<i32>` in constants
##[error]  --> coretests/tests/bool.rs:96:34
   |
96 |     const A: Option<i32> = false.then_some(0);
   |                                  ^^^^^^^^^^^^
   |
   = note: calls in constants are limited to constant functions, tuple structs and tuple variants

error[E0015]: cannot call non-const method `core::bool::<impl bool>::then_some::<i32>` in constants
##[error]  --> coretests/tests/bool.rs:97:33
   |
97 |     const B: Option<i32> = true.then_some(0);
   |                                 ^^^^^^^^^^^^
   |
   = note: calls in constants are limited to constant functions, tuple structs and tuple variants

error[E0015]: cannot call non-const method `core::bool::<impl bool>::then::<i32, fn() -> i32 {bool::test_bool_to_option::zero}>` in constants
##[error]  --> coretests/tests/bool.rs:98:34
   |
98 |     const C: Option<i32> = false.then(zero);
   |                                  ^^^^^^^^^^
   |
   = note: calls in constants are limited to constant functions, tuple structs and tuple variants

error[E0015]: cannot call non-const method `core::bool::<impl bool>::then::<i32, fn() -> i32 {bool::test_bool_to_option::zero}>` in constants
##[error]  --> coretests/tests/bool.rs:99:33
   |
99 |     const D: Option<i32> = true.then(zero);
   |                                 ^^^^^^^^^^
   |
   = note: calls in constants are limited to constant functions, tuple structs and tuple variants

For more information about this error, try `rustc --explain E0015`.
[RUSTC-TIMING] coretests test:true 59.426
error: could not compile `coretests` (test "coretests") due to 4 previous errors
env -u RUSTC_WRAPPER CARGO_ENCODED_RUSTDOCFLAGS="-Csymbol-mangling-version=v0\u{1f}-Zrandomize-layout\u{1f}-Zunstable-options\u{1f}--check-cfg=cfg(bootstrap)\u{1f}-Dwarnings\u{1f}-Wrustdoc::invalid_codeblock_attributes\u{1f}--crate-version\u{1f}1.92.0-nightly\t(739d88487\t2025-10-03)" CARGO_ENCODED_RUSTFLAGS="-Csymbol-mangling-version=v0\u{1f}-Zrandomize-layout\u{1f}-Zunstable-options\u{1f}--check-cfg=cfg(bootstrap)\u{1f}-Zmacro-backtrace\u{1f}-Csplit-debuginfo=off\u{1f}-Clink-arg=-L/usr/lib/llvm-20/lib\u{1f}-Cllvm-args=-import-instr-limit=10\u{1f}-Clink-args=-Wl,-z,origin\u{1f}-Clink-args=-Wl,-rpath,$ORIGIN/../lib\u{1f}-Alinker-messages\u{1f}--cap-lints=allow\u{1f}--cfg\u{1f}randomized_layouts" RUSTC="/checkout/obj/build/aarch64-unknown-linux-gnu/stage2-codegen/cg_clif/dist/rustc-clif" RUSTDOC="/checkout/obj/build/aarch64-unknown-linux-gnu/stage2-codegen/cg_clif/dist/rustdoc-clif" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage0/bin/cargo" "test" "--manifest-path" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2-codegen/cg_clif/build/sysroot_tests/Cargo.toml" "--target-dir" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2-codegen/cg_clif/build/sysroot_tests_target" "--locked" "--target" "aarch64-unknown-linux-gnu" "-p" "coretests" "-p" "alloctests" "--tests" "--" "-q" exited with status ExitStatus(unix_wait_status(25856))
Command `/checkout/obj/build/aarch64-unknown-linux-gnu/stage0/bin/cargo run --target aarch64-unknown-linux-gnu -Zbinary-dep-depinfo -j 4 -Zroot-dir=/checkout --locked --color always --release --manifest-path /checkout/compiler/rustc_codegen_cranelift/build_system/Cargo.toml -- test --download-dir /checkout/obj/build/cg_clif_download --out-dir /checkout/obj/build/aarch64-unknown-linux-gnu/stage2-codegen/cg_clif --no-unstable-features --use-backend cranelift --sysroot llvm --skip-test testsuite.extended_sysroot [workdir=/checkout/compiler/rustc_codegen_cranelift]` failed with exit code 1
Created at: src/bootstrap/src/core/build_steps/test.rs:3658:25
Executed at: src/bootstrap/src/core/build_steps/test.rs:3703:26

Command has failed. Rerun with -v to see more details.
Bootstrap failed while executing `--stage 2 test --skip tests --skip coverage-map --skip coverage-run --skip library --skip tidyselftest`
Build completed unsuccessfully in 0:22:01
  local time: Fri Oct  3 17:21:07 UTC 2025
  network time: Fri, 03 Oct 2025 17:21:07 GMT
##[error]Process completed with exit code 1.
```

</details>

rust-log-analyzer commented:
The job **`tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/18230004675/job/51910916700) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/51910916700) [(plain)](https://github.com/rust-lang/rust/commit/2d11909e33c9bfe3f7c25729a388f2ee8ce34280/checks/51910916700/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
[TIMING:end] tool::Tidy { compiler: Compiler { stage: 0, host: x86_64-unknown-linux-gnu, forced_compiler: false }, target: x86_64-unknown-linux-gnu } -- 0.000
fmt check
Diff in /checkout/library/core/src/bool.rs:60:
     #[inline]
     pub const fn then<T, F: FnOnce() -> T>(self, f: F) -> Option<T>
     where
-        F: ~const FnOnce() -> T,
+        F: [const] FnOnce() -> T,
     {
         if self { Some(f()) } else { None }
     }
fmt: checked 6451 files
Bootstrap failed while executing `test src/tools/tidy tidyselftest --extra-checks=py,cpp,js,spellcheck`
Build completed unsuccessfully in 0:00:48
  local time: Fri Oct  3 18:08:09 UTC 2025
```

</details>

dollaransh17 committed:
Add missing const stability attributes for bool methods

This fixes compilation errors in bootstrap tests by adding:
- rustc_const_stable attribute for then_some method
- rustc_const_unstable attribute for then method
- [const] Destruct bound for const trait compatibility
- Required Destruct import

Resolves E0493 destructor evaluation errors and missing
const stability attribute errors during bootstrap.

rust-log-analyzer commented:
The job **`aarch64-gnu-llvm-20-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/18231104202/job/51914451772) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/51914451772) [(plain)](https://github.com/rust-lang/rust/commit/7e6ad283995aff3284b616c0b55f56715b4361cf/checks/51914451772/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
[RUSTC-TIMING] build_script_build test:false 0.339
error[E0493]: destructor of `T` cannot be evaluated at compile-time
  --> library/core/src/bool.rs:36:37
   |
36 |     pub const fn then_some<T>(self, t: T) -> Option<T> {
   |                                     ^ the destructor for this type cannot be evaluated in constant functions
37 |         if self { Some(t) } else { None }
38 |     }
   |     - value is dropped here

For more information about this error, try `rustc --explain E0493`.
[RUSTC-TIMING] core test:false 24.092
```

</details>
